asio/include/asio/error_as_deferred.hpp

```diff
@@ -0,0 +1,147 @@
+//
+// error_as_deferred.hpp
+// ~~~~~~~~~~~~
+//
+// Copyright (c) 2003-2022 Christopher M. Kohlhoff (chris at kohlhoff dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#ifndef ASIO_ERROR_DEFERRED_HPP
+#define ASIO_ERROR_DEFERRED_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+# pragma once
+#endif // defined(_MSC_VER) && (_MSC_VER >= 1200)
+
+
+#include "asio/deferred.hpp"
+#include "asio/error.hpp"
+
+#include "asio/detail/config.hpp"
+
+#if (defined(ASIO_HAS_STD_TUPLE) \
+    && defined(ASIO_HAS_DECLTYPE) \
+    && defined(ASIO_HAS_VARIADIC_TEMPLATES)) \
+  || defined(GENERATING_DOCUMENTATION)
+
+#include <tuple>
+#include "asio/associator.hpp"
+#include "asio/async_result.hpp"
+#include "asio/detail/type_traits.hpp"
+#include "asio/detail/utility.hpp"
+
+#include "asio/detail/push_options.hpp"
+
+namespace asio {
+
+
+template <typename Signature>
+class async_result<asio::error_code, Signature>
+{
+public:
+
+  struct deferred_intercept_error_code
+  {
+    asio::error_code &ec;
+
+    template<typename ... Args>
+    auto operator()(asio::error_code ec_, Args && ... args) const
+      -> deferred_values<Args...>
+    {
+      ec = ec_;
+      return asio::deferred_t::values(std::forward<Args>(args)...);
```

Comment:
Would there be performance overhead in using this completion mechanism over, say, directly using `redirect_error(use_awaitable,ec)`?